### ISSUE TITLE
AMP Review 1

### ISSUE THREAD

**Issue by andrewrreed** [fastforwardlabs/question_answering#2]
# Purpose

The purpose of this issue is to track notes / edits / suggestions while reviewing the AMP.

# Comments / Feedback / Suggestions

- [x] 1. I know its a pain, but all of these should probably say 2021 not 2020 :/
https://github.com/fastforwardlabs/question_answering/blob/e6943909b49a3ff3b7aa0ad5578471d2499ef909/apps/data_visualizer.py#L4
- [x] 2. Lets make this "Deep Learning for Question Answering" to follow suit with the H1 title in the README. (I also updated the AMP Title in the tracking sheet to reflect this...feel free to push back if you feel strongly otherwise)
https://github.com/fastforwardlabs/question_answering/blob/e6943909b49a3ff3b7aa0ad5578471d2499ef909/.project-metadata.yaml#L1
- [x] 3. As discussed via Zoom the other day, we could improve the focus and usability of the AMP by logically separating out the model training info/code from the UI apps that seem to be the main focus of the AMP (at least at launch):
  - [x] a. Move all README content from "For NLP Practitioners" and below into its own README.md in the `qa` (or `scripts`?) directory
  - [x] b. Tailor this new README content towards the specific task: "Hey, here's how you can train your own QA model given some data... then you can evaluate it and use it in the fancy Streamlit apps"
  - [x] c. Include a quick discussion on how to use the `launch_tensorboard` app (or maybe want to remove that functionality if it doesn't actually get referenced or mentioned anywhere...)
  - [x] d. Reference and link to the new README from somewhere in the old README
- [x] 4. Super minor, but looks like the top line formatting of the `LICENSE.txt` file got jumbled
https://github.com/fastforwardlabs/question_answering/blob/e6943909b49a3ff3b7aa0ad5578471d2499ef909/LICENSE.txt#L1
- [x] 5. Comments on the `README.md`:
  - [x] CML has a bit of a wonky markdown renderer, so this HTML syntax doesn't work for any of the images you have in the `README.md`. The syntax that should work for you is `![WikiQA question answering interface](images/Screenshot_ModelExplorer.png)`... however, you'll likely need to resize the images manually since you can't do it inline 😒  (here's [a quick way to do it](https://www.businessinsider.com/how-to-resize-an-image-on-mac#:~:text=Select%20your%20photo%20or%20photos%20by%20highlighting%20them%20on%20the,and%20then%20%22Adjust%20size.%22) using the Preview App on Mac)
![image](https://user-images.githubusercontent.com/23462538/105186063-4e321f00-5aff-11eb-9ac3-08e6d8c4606d.png)

**Comment by andrewrreed** [comment]
In response to the changes implemented to address the comments above, here are a few followups (numbering stays consistent with above):
- [x] 1.  ❌   Looks like we've got one straggler
https://github.com/fastforwardlabs/question_answering/blob/f39d582a15724a3e83a21982133e16bca1f53e43/setup.py#L4
- [x] 3.c I don't think I see any discussion of tensorboard in the new `scripts/README.md`. If `apps/launch_tensorboard.py` already points to the right directory that the model training checkpoints save to, lets include a quick note that you can launch that as an app to monitor training. If its not already setup and is more work, lets maybe get rid of it as to not confuse people?
- [x] 3.d Might be a bit petty, but lets link to `scripts/README.md` from the mention in the main README.

**Comment by andrewrreed** [comment]
# Review Comments cont.

- [x] 6. In the README, towards the bottom, can we include specific instructions / options for running on CML. This is a new requirement that Alex B just asked for today after seeing how it's done in the NeuralQA and Anomaly Detection AMPS (I just updated the STS repo to have it as well, so that might be a good place for you to check since this README is similar to that one). But the general idea is to have something like this:

## Deploying on CML

There are three ways to launch this project on CML:

1. **From Prototype Catalog** - Navigate to the Prototype Catalog on a CML workspace, select the "Deep Learning for Question Answering" tile, click "Launch as Project", click "Configure Project"
2. **As ML Prototype** - In a CML workspace, click "New Project", add a Project Name, select "ML Prototype" as the Initial Setup option, copy in the [repo URL](), click "Create Project", click "Configure Project"
3. **Manual Setup** - In a CML workspace, click "New Project", add a Project Name, select "Git" as the Initial Setup option, copy in the [repo URL](), click "Create Project". Launch a Python3 Workbench Session with at least 4GB of memory and 2vCPUs. Then follow these instructions below, in order.

**Comment by andrewrreed** [comment]
- [x] 7. The "script" link in `scripts/README.md` appears broken:
https://github.com/fastforwardlabs/question_answering/blame/f39d582a15724a3e83a21982133e16bca1f53e43/scripts/README.md#L5
- [x] 8. Can you follow the instructions in [this doc](https://docs.google.com/document/d/1Ru0aVCDiUnkL4IDDZV8LzAorKjvyeQtWeTOWMx007-Y/edit#heading=h.83u4ytyppmgc) to document the files/lines that are borrowed. That way I can go through to check that they have inline code comments to reference HuggingFace's source code + we can include in the Jira ticket so Legal generates the proper NOTICES file for us.
- [x] 9. Can you screen grab whatever thumbnail image you want to use as the visual in the CML Prototype Catalog?

**Comment by andrewrreed** [comment]
@melaniebeck FYI - I moved the PR content to this issue and added a few more review items. I'm also tracking overall status in the Review doc [here](https://docs.google.com/document/d/1c7RC6ByVB8A3CCG2knyf3rqsWjEOZOxWQ1UpjoZfI50/edit#).

Still waiting for the Eng bug fix to cross off the general functionality tests. Other than that, I'm not really sure how to test the model training bits because I don't have access to an environment with enough RAM. Were you able to test this? If so, how?

**Comment by melaniebeck** [comment]
I think I've addressed the issues above and it's ready for another round of your careful eye. 

As for testing, I have tested all other functionality on my CDSW_QuestionAnswering project on fflab-4 by running the scripts as I described in the README. Most of them were run in Sessions, some ran as Experiments, and some as Jobs. I have also successfully deployed a Model and pinged it from my laptop via the command line.  Some of this testing was with very small datasets and without GPUs (since I assumed most people wouldn't have access).  However, I have also tested most (all?) of them with GPUs as well --  just yesterday I tested the `multi_train.py` script as a Job using the GPU, and it worked just fine.
